### ISSUE TITLE
Generic type erased after type guard with strictNullChecks

### ISSUE THREAD

**Issue by jgoz** [Microsoft/TypeScript#10811]
**TypeScript Version:**  2.0.2

**Code**

```ts
interface Pet {
    name: string;
}

function isPet(pet: any): pet is Pet {
    return typeof pet.name === "string";
}

export function speak<TPet extends Pet>(pet: TPet, voice: (pet: TPet) => string): string {
    if (!isPet(pet)) {
        throw new Error("Expected \"pet\" to be a Pet");
    }

    return voice(pet); // line 14
}
```

**Expected behavior:**
Module compiles without error when `--strictNullChecks` is enabled OR disabled.

**Actual behavior:** 
- Module compiles without error when `--strictNullChecks` is _disabled_.
- The following error is produced when `--strictNullChecks` is _enabled_

    ```
    bug.ts(14,18): error TS2345: Argument of type 'Pet' is not assignable to parameter of type 'TPet'.
    ```

**Comment by ahejlsberg** [comment]
Issue here is that in the true branch of `isPet(pet)`, we should keep type `TPet extends Pet` for `pet` instead of narrowing to type `Pet` because the former is a more specific type. It's an easy fix, I will put up a PR soon. BTW, this doesn't appear to have anything to do with `--strictNullChecks`, the error is reported in either mode.
